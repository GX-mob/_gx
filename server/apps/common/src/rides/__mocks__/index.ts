import { PriceDetail, Price } from "@app/database";
import { geometry } from "@app/helpers";
//@ts-ignore
const { decode } = require("google-polyline");

export const rideType1: PriceDetail = {
  type: 1,
  available: true,
  perKilometer: 1.1,
  perMinute: 0.3,
  kilometerMultipler: 0.2,
  minuteMultipler: 0.1,
  overBusinessTimeKmAdd: 0.4,
  overBusinessTimeMinuteAdd: 0.3,
};

export const rideType2: PriceDetail = {
  type: 2,
  available: true,
  perKilometer: 1.6,
  perMinute: 0.5,
  kilometerMultipler: 0.3,
  minuteMultipler: 0.2,
  overBusinessTimeKmAdd: 0.6,
  overBusinessTimeMinuteAdd: 0.5,
};

export const prices: Price[] = [
  {
    area: "AL",
    currency: "BRL",
    timezone: "America/Maceio",
    general: [rideType1, rideType2],
    subAreas: {
      maceio: [rideType1, rideType2],
    },
  },
  {
    area: "PE",
    currency: "BRL",
    timezone: "America/Maceio",
    general: [rideType1, rideType2],
    subAreas: {
      recife: [rideType1, rideType2],
    },
  },
];

export const pathEncodedMock =
  "_jn~Fh_}uOlIr@dNxCxIOxIgB|HmElEmE~BeI~BsHjAwIh@yHjAkLdDcHxCkDjCwBfFcApCIdDO~B?dDc@dD?";
export const pathDecodedMock = decode(pathEncodedMock);
export const pathDistance = geometry.distance.meterToKM(
  geometry.distance.path(pathDecodedMock),
);
export const pathDuration = 10;

export const path = {
  encoded: pathEncodedMock,
  decoded: pathDecodedMock,
  distance: pathDistance,
  duration: pathDuration,
};
