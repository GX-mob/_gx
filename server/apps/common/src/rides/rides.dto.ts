import {
  Ride,
  RideTypes,
  RidePayMethods,
  TRoute,
  TRoutePoint,
} from "@app/database";
import {
  ValidateNested,
  IsNotEmpty,
  IsEnum,
  IsInt,
  IsArray,
} from "class-validator";

export class GetRidesPricesParams {
  area!: string;
  subArea?: string;
}

export class CreateRideDto {
  @ValidateNested()
  route!: Route;

  @IsEnum(RideTypes)
  type!: RideTypes;

  @IsEnum(RidePayMethods)
  payMethod!: RidePayMethods;

  @IsNotEmpty()
  country!: Ride["country"];

  @IsNotEmpty()
  area!: Ride["area"];

  @IsNotEmpty()
  subArea!: Ride["subArea"];
}

class Route implements TRoute {
  @ValidateNested()
  start!: Point;

  @ValidateNested()
  waypoints?: Point[];

  @ValidateNested()
  end!: Point;

  @IsNotEmpty()
  path!: string;

  @IsInt()
  distance!: number;

  @IsInt()
  duration!: number;
}

class Point implements TRoutePoint {
  @IsArray()
  coord!: [number, number];

  @IsNotEmpty()
  primary!: string;

  @IsNotEmpty()
  secondary!: string;

  @IsNotEmpty()
  district!: string;
}
