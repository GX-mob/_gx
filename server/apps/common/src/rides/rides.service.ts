import { Injectable } from "@nestjs/common";
import { DatabaseService, Price, PriceDetail, Ride } from "@app/database";
import { util } from "@app/helpers";

@Injectable()
export class RidesService {
  private areas: { [area: string]: Price } = {};

  constructor(readonly database: DatabaseService) {
    this.init();
  }

  private async init() {
    const priceModel = this.database.priceModel;

    /**
     * Get and store all rides types and prices
     */
    const prices = await priceModel.find().lean();

    if (!prices.length) {
      throw new Error("Empty rides types list");
    }
    prices.forEach((price) => {
      this.areas[price.area] = price;
    });

    /**
     * Watch prices update
     */
    priceModel.watch().on("change", (data) => {
      switch (data.operationType) {
        case "update":
        case "insert":
          const { fullDocument } = data;
          this.areas[fullDocument.area] = fullDocument;
          break;
        case "delete":
          delete this.areas[fullDocument.area];
          break;
      }
    });
  }

  /**
   * Returns the rides types and respective prices of the requested area
   * or the price for the respective area for the ride type, if provided a type
   * fallback to $area if not have results for $area.$subArea
   * @param area
   * @param subArea
   * @returns {PriceDetail[] | PriceDetail | undefined} Price list, price for provided ride type or undefined
   */
  getPrice(
    area: string,
    subArea?: string,
    rideType?: Ride["type"],
  ): PriceDetail[] | PriceDetail | undefined {
    if (!util.hasProp(this.areas, area) || !this.areas[area]) {
      return;
    }

    const areaPrices = this.areas[area];

    const response =
      subArea && areaPrices.subAreas[subArea]
        ? areaPrices.subAreas[subArea]
        : areaPrices.general;

    if (!response) return;

    if (typeof rideType !== "undefined") {
      return response.find((price) => price.type === rideType);
    }

    return response;
  }
}
