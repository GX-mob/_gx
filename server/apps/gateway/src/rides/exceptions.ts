import {
  UnprocessableEntityException,
  NotFoundException,
  ForbiddenException,
} from "@nestjs/common";

export class RideNotFoundException extends NotFoundException {}
export class InvalidRideTypeException extends UnprocessableEntityException {}
export class UnsupportedAreaException extends UnprocessableEntityException {}
export class RideNoReadPermission extends ForbiddenException {}
