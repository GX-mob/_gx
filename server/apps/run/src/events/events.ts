import {
  Configuration,
  configurationSchema,
  //
  DriverRideAcceptedResponse,
  driverRideAcceptedResponseSchema,
  //
  VoyagerRideAcceptedResponse,
  voyagerRideAcceptedResponseSchema,
  //
  OfferRequest,
  offerSchema,
  //
  OfferResponse,
  offerReponseSchema,
  //
  OfferSent,
  offerSentSchema,
  //
  Position,
  positionSchema,
  //
  Setup,
  driverSetupSchema,
  //
  State,
  stateSchema,
  //
  CancelRide,
  cancelRideSchema,
  //
  CanceledRide,
  canceledRideSchema,
  //
  delayedOfferReponse,
  DelayedOfferReponse,
  //
  AmIRunning,
  amIRunningSchema,
  //
  OfferGotTooLong,
  offerGotTooLong,
} from "./schemas";

export * from "./schemas";
export * from "./schemas/common";

export enum EVENTS {
  CONFIGURATION = "CONFIGURATION",
  DRIVER_RIDE_ACCEPTED_RESPONSE = "DRIVER_RIDE_ACCEPTED_RESPONSE",
  VOYAGER_RIDE_ACCEPTED_RESPONSE = "VOYAGER_RIDE_ACCEPTED_RESPONSE",
  OFFER = "OFFER",
  OFFER_RESPONSE = "OFFER_RESPONSE",
  OFFER_SENT = "OFFER_SENT",
  POSITION = "POSITION",
  DRIVER_SETUP = "DRIVER_SETUP",
  STATE = "STATE",
  AM_I_RUNNING = "AM_I_RUNNING",
  CANCEL_RIDE = "CANCEL_RIDE",
  CANCELED_RIDE = "CANCELED_RIDE",
  DELAYED_OFFER_RESPONSE = "DELAYED_OFFER_RESPONSE",
  OFFER_GOT_TOO_LONG = "OFFER_GOT_TOO_LONG",
}

export const serverEventsSchemas = {
  [EVENTS.CONFIGURATION]: {
    id: 1,
    schema: configurationSchema,
  },
  [EVENTS.DRIVER_RIDE_ACCEPTED_RESPONSE]: {
    id: 2,
    schema: driverRideAcceptedResponseSchema,
  },
  [EVENTS.VOYAGER_RIDE_ACCEPTED_RESPONSE]: {
    id: 3,
    schema: voyagerRideAcceptedResponseSchema,
  },
  [EVENTS.OFFER]: {
    id: 4,
    schema: offerSchema,
  },
  [EVENTS.OFFER_RESPONSE]: {
    id: 5,
    schema: offerReponseSchema,
  },
  [EVENTS.OFFER_SENT]: {
    id: 6,
    schema: offerSentSchema,
  },
  [EVENTS.POSITION]: {
    id: 7,
    schema: positionSchema,
  },
  [EVENTS.DRIVER_SETUP]: {
    id: 8,
    schema: driverSetupSchema,
  },
  [EVENTS.STATE]: {
    id: 9,
    schema: stateSchema,
  },
  [EVENTS.AM_I_RUNNING]: {
    id: 10,
    schema: amIRunningSchema,
  },
  [EVENTS.CANCEL_RIDE]: {
    id: 11,
    schema: cancelRideSchema,
  },
  [EVENTS.CANCELED_RIDE]: {
    id: 12,
    schema: canceledRideSchema,
  },
  [EVENTS.DELAYED_OFFER_RESPONSE]: {
    id: 13,
    schema: delayedOfferReponse,
  },
  [EVENTS.OFFER_GOT_TOO_LONG]: {
    id: 14,
    schema: offerGotTooLong,
  },
};

export interface Events {
  [EVENTS.CONFIGURATION]: Configuration;
  [EVENTS.DRIVER_RIDE_ACCEPTED_RESPONSE]: DriverRideAcceptedResponse;
  [EVENTS.VOYAGER_RIDE_ACCEPTED_RESPONSE]: VoyagerRideAcceptedResponse;
  [EVENTS.OFFER]: OfferRequest;
  [EVENTS.OFFER_RESPONSE]: OfferResponse;
  [EVENTS.OFFER_SENT]: OfferSent;
  [EVENTS.POSITION]: Position;
  [EVENTS.DRIVER_SETUP]: Setup;
  [EVENTS.STATE]: State;
  [EVENTS.AM_I_RUNNING]: AmIRunning;
  [EVENTS.CANCEL_RIDE]: CancelRide;
  [EVENTS.CANCELED_RIDE]: CanceledRide;
  [EVENTS.DELAYED_OFFER_RESPONSE]: DelayedOfferReponse;
  [EVENTS.OFFER_GOT_TOO_LONG]: OfferGotTooLong;
}
