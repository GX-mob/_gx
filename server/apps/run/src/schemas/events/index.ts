export { default as configuration } from "./configuration";
export { default as driverOfferAccepted } from "./driver-offer-accepted";
export { default as offer } from "./offer";
export { default as position } from "./position";
export { default as offerReponse } from "./offer-response";
export { default as offerSent } from "./offer-sent";
export { default as setup } from "./setup";
export { default as state } from "./state";
