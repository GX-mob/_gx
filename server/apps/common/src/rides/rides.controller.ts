import { Controller, UseGuards, Get } from "@nestjs/common";
import { AuthGuard, AuthorizedRequest } from "@app/auth";
import { DataService } from "@app/data";
import { RidesService } from "./rides.service";

@Controller("rides/")
@UseGuards(AuthGuard)
export class RidesController {
  constructor(readonly data: DataService, readonly rideService: RidesService) {}
}
