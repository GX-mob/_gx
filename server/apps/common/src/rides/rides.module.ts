import { Module } from "@nestjs/common";
import { CacheModule } from "@app/cache";
import { DataModule } from "@app/data";
import { DatabaseModule } from "@app/database";
import { SessionModule } from "@app/session";
import { RidesService } from "./rides.service";
import { RidesController } from "./rides.controller";

@Module({
  imports: [CacheModule, DatabaseModule, DataModule, SessionModule],
  providers: [RidesService],
  controllers: [RidesController],
})
export class RidesModule {}
