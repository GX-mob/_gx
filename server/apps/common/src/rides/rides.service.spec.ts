/**
 * Data Service
 *
 * @group unit/services/ride
 */
import { Test, TestingModule } from "@nestjs/testing";
import { parseISO } from "date-fns";
import {
  DatabaseModule,
  DatabaseService,
  Price,
  PriceDetail,
  TRoutePoint,
  RideTypes,
} from "@app/database";
import { RidesService } from "./rides.service";
import { EventEmitter } from "events";
import { geometry } from "@app/helpers";
import { utcToZonedTime } from "date-fns-tz";
//@ts-ignore
const { decode } = require("google-polyline");

describe("RideService", () => {
  let service: RidesService;
  const emitter = new EventEmitter();

  emitter.setMaxListeners(50);

  const rideType1: PriceDetail = {
    type: 1,
    available: true,
    perKilometer: 1.1,
    perMinute: 0.3,
    kilometerMultipler: 0.2,
    minuteMultipler: 0.1,
    overBusinessTimeKmAdd: 0.4,
    overBusinessTimeMinuteAdd: 0.3,
  };

  const rideType2: PriceDetail = {
    type: 2,
    available: true,
    perKilometer: 1.6,
    perMinute: 0.5,
    kilometerMultipler: 0.3,
    minuteMultipler: 0.2,
    overBusinessTimeKmAdd: 0.6,
    overBusinessTimeMinuteAdd: 0.5,
  };

  const prices: Price[] = [
    {
      area: "AL",
      currency: "BRL",
      timezone: "America/Maceio",
      general: [rideType1, rideType2],
      subAreas: {
        maceio: [rideType1, rideType2],
      },
    },
    {
      area: "PE",
      currency: "BRL",
      timezone: "America/Maceio",
      general: [rideType1, rideType2],
      subAreas: {
        recife: [rideType1, rideType2],
      },
    },
  ];

  const databaseServiceMock = {
    priceModel: {
      find: () => ({
        lean: async (): Promise<Price[]> => [...prices],
      }),
      watch: () => emitter,
    },
  };

  const pathEncodedMock =
    "_jn~Fh_}uOlIr@dNxCxIOxIgB|HmElEmE~BeI~BsHjAwIh@yHjAkLdDcHxCkDjCwBfFcApCIdDO~B?dDc@dD?";
  const pathDecodedMock = decode(pathEncodedMock);
  const pathDistance = geometry.distance.meterToKM(
    geometry.distance.path(pathDecodedMock),
  );
  const pathDuration = 10;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [DatabaseModule],
      providers: [RidesService],
    })
      .overrideProvider(DatabaseService)
      .useValue(databaseServiceMock)
      .compile();

    service = module.get<RidesService>(RidesService);
  });

  it("should be defined", () => {
    expect(service).toBeDefined();
  });

  it("should has rides types loaded", async () => {
    const prices = await databaseServiceMock.priceModel.find().lean();
    const areas: { [area: string]: Price } = {};

    prices.forEach((price) => {
      areas[price.area] = price;
    });

    expect(service.areas).toStrictEqual(areas);
  });

  it("should update area prices", async () => {
    const prices = await databaseServiceMock.priceModel.find().lean();

    const newRideType1Values = {
      ...rideType1,
      perKilometer: 1.7,
      perMinute: 0.6,
      kilometerMultipler: 0.4,
      minuteMultipler: 0.3,
      overBusinessTimeKmAdd: 0.7,
      overBusinessTimeMinuteAdd: 0.6,
    };

    emitter.emit("change", {
      operationType: "update",
      fullDocument: {
        ...prices[0],
        general: [prices[0].general[1], newRideType1Values],
      },
    });

    const general = service.areas[prices[0].area].general;

    expect(general).toStrictEqual([prices[0].general[1], newRideType1Values]);
  });

  it("should insert area prices", async () => {
    const newRideType1Values = {
      ...rideType1,
      perKilometer: 1.7,
      perMinute: 0.6,
      kilometerMultipler: 0.4,
      minuteMultipler: 0.3,
      overBusinessTimeKmAdd: 0.7,
      overBusinessTimeMinuteAdd: 0.6,
    };

    const newDocument = {
      ...prices[0],
      area: "SP",
      general: [prices[0].general[1], newRideType1Values],
      subAreas: {},
    };

    emitter.emit("change", {
      operationType: "insert",
      fullDocument: newDocument,
    });

    const general = service.areas[newDocument.area];

    expect(general).toStrictEqual(newDocument);
  });

  describe("getPrice", () => {
    it("should return undefined", () => {
      expect(service.getRideStatusPrice("MG")).toBe(undefined);
    });

    it("should return the price list of an area", () => {
      expect(service.getRideStatusPrice("AL")).toStrictEqual(prices[0].general);
    });

    it("should return the price list of a subArea", () => {
      expect(service.getRideStatusPrice("AL", "maceio")).toStrictEqual(
        prices[0].subAreas["maceio"],
      );
    });

    it("should return the fallback to area due to an undefined subArea", () => {
      expect(service.getRideStatusPrice("AL", "arapiraca")).toStrictEqual(
        prices[0].general,
      );
    });

    it("should return the price of a ride type", () => {
      const [, ridePrice] = prices[0].subAreas["maceio"];

      expect(
        service.getRideStatusPrice("AL", "maceio", ridePrice.type),
      ).toStrictEqual(ridePrice);
    });

    it("should return the fallback price of a ride type", () => {
      const [, ridePrice] = prices[0].subAreas["maceio"];

      expect(
        service.getRideStatusPrice("AL", "arapiraca", ridePrice.type),
      ).toStrictEqual(ridePrice);
    });
  });

  describe("isBusinessTime", () => {
    it("should return true", () => {
      const dateTime = parseISO("2020-06-13 12:00");
      const dateTime2 = parseISO("2020-06-13 20:00");

      expect(service.isBusinessTime("America/Maceio", dateTime)).toBeTruthy();
      expect(service.isBusinessTime("America/Maceio", dateTime2)).toBeTruthy();
    });

    it("should return false", () => {
      const dateTime = parseISO("2020-06-13 11:00");
      const dateTime2 = parseISO("2020-06-13 22:00");
      const sunday = parseISO("2020-06-14 12:00");

      expect(service.isBusinessTime("America/Maceio", dateTime)).toBeFalsy();
      expect(service.isBusinessTime("America/Maceio", dateTime2)).toBeFalsy();
      expect(service.isBusinessTime("America/Maceio", sunday)).toBeFalsy();
    });
  });

  function priceCalculationTest(
    func: "distancePrice" | "durationPrice",
    value: number,
  ) {
    describe(func, () => {
      it("should calculate", () => {
        const costsType1InBusiness = service[func](
          value,
          prices[0].general[0],
          true,
        );
        const costsType2InBusiness = service[func](
          value,
          prices[0].general[1],
          true,
        );
        const costsType1OutBusiness = service[func](
          value,
          prices[0].general[0],
          false,
        );
        const costsType2OutBusiness = service[func](
          value,
          prices[0].general[1],
          false,
        );

        expect(costsType1InBusiness).toMatchSnapshot();
        expect(costsType2InBusiness).toMatchSnapshot();

        expect(costsType1OutBusiness).toMatchSnapshot();
        expect(costsType2OutBusiness).toMatchSnapshot();

        expect(
          costsType1InBusiness.total < costsType1OutBusiness.total,
        ).toBeTruthy();
        expect(
          costsType2InBusiness.total < costsType2OutBusiness.total,
        ).toBeTruthy();
      });
    });
  }

  priceCalculationTest("distancePrice", pathDistance);
  priceCalculationTest("durationPrice", pathDuration);

  describe("getRideCosts", () => {
    it("should calculate ", () => {
      const point: TRoutePoint = {
        coord: [0, 0],
        primary: "foo",
        secondary: "foo",
        district: "foo",
      };

      const mockCreateRideDto: any = {};
      mockCreateRideDto.route = {
        start: point,
        end: point,
        path: pathEncodedMock,
        distance: pathDistance,
        duration: pathDuration,
      };
      mockCreateRideDto.area = "AL";
      mockCreateRideDto.subArea = "maceio";
      mockCreateRideDto.type = RideTypes.Normal;

      const distancePrice = service.distancePrice(
        pathDistance,
        prices[0].general[0],
        false,
      );
      const durationPrice = service.durationPrice(
        pathDuration,
        prices[0].general[0],
        false,
      );

      const total = distancePrice.total + durationPrice.total;

      const ridePrice = service.getRideCosts(mockCreateRideDto);

      expect(distancePrice).toStrictEqual(ridePrice.distance);
      expect(durationPrice).toStrictEqual(ridePrice.duration);
    });
  });
});
