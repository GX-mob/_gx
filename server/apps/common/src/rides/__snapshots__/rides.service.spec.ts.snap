// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RideService distancePrice should calculate 1`] = `
Object {
  "aditionalForLongRide": 0,
  "aditionalForOutBusinessTime": 0,
  "total": 2,
}
`;

exports[`RideService distancePrice should calculate 2`] = `
Object {
  "aditionalForLongRide": 0,
  "aditionalForOutBusinessTime": 0,
  "total": 2.8,
}
`;

exports[`RideService distancePrice should calculate 3`] = `
Object {
  "aditionalForLongRide": 0,
  "aditionalForOutBusinessTime": 0.7,
  "total": 2.7,
}
`;

exports[`RideService distancePrice should calculate 4`] = `
Object {
  "aditionalForLongRide": 0,
  "aditionalForOutBusinessTime": 1.1,
  "total": 3.9,
}
`;

exports[`RideService durationPrice should calculate 1`] = `
Object {
  "aditionalForLongRide": 0,
  "aditionalForOutBusinessTime": 0,
  "total": 3,
}
`;

exports[`RideService durationPrice should calculate 2`] = `
Object {
  "aditionalForLongRide": 0,
  "aditionalForOutBusinessTime": 0,
  "total": 5,
}
`;

exports[`RideService durationPrice should calculate 3`] = `
Object {
  "aditionalForLongRide": 0,
  "aditionalForOutBusinessTime": 1,
  "total": 6,
}
`;

exports[`RideService durationPrice should calculate 4`] = `
Object {
  "aditionalForLongRide": 0,
  "aditionalForOutBusinessTime": 2,
  "total": 10,
}
`;
